cmd/rekor_monitor/main.go

```diff
@@ -173,116 +127,71 @@ func mainLoopV1(serverURL string, once bool, logInfoFile string, interval time.D
-		if len(certID.Issuers) == 0 {
-			fmt.Printf("Monitoring certificate subject %s\n", certID.CertSubject)
-		} else {
-			fmt.Printf("Monitoring certificate subject %s for issuer(s) %s\n", certID.CertSubject, strings.Join(certID.Issuers, ","))
-		}
-	}
-	for _, fp := range monitoredValues.Fingerprints {
-		fmt.Printf("Monitoring fingerprint %s\n", fp)
-	}
-	for _, sub := range monitoredValues.Subjects {
-		fmt.Printf("Monitoring subject %s\n", sub)
-	}
-
-	ticker := time.NewTicker(interval)
-	defer ticker.Stop()
-
-	ctx, stop := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
-	defer stop()
-
-	// To get an immediate first tick, for-select is at the end of the loop
-	for {
-		inputEndIndex := config.EndIndex
-
-		// TODO: Handle Rekor sharding
-		// https://github.com/sigstore/rekor-monitor/issues/57
-		var logInfo *models.LogInfo
-		var prevCheckpoint *util.SignedCheckpoint
-		prevCheckpoint, logInfo, err = rekor_v1.RunConsistencyCheck(rekorClient, verifier, logInfoFile)
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "error running consistency check: %v", err)
-			return
-		}
-
-		if config.StartIndex == nil {
-			if prevCheckpoint != nil {
-				checkpointStartIndex := rekor_v1.GetCheckpointIndex(logInfo, prevCheckpoint)
-				config.StartIndex = &checkpointStartIndex
-			} else {
-				fmt.Fprintf(os.Stderr, "no start index set and no log checkpoint")
-				return
-			}
-		}
-
-		if config.EndIndex == nil {
-			checkpoint, err := rekor_v1.ReadLatestCheckpoint(logInfo)
+	cmd.PrintMonitoredValues(monitoredValues)
+	cmd.LoopLogs(cmd.LoopLogsParams{
+		Interval:        flags.Interval,
+		Config:          config,
+		MonitoredValues: monitoredValues,
+		Once:            flags.Once,
+		RunConsistencyCheckFn: func(_ context.Context) (cmd.Checkpoint, cmd.Checkpoint, error) {
+			prev, cur, err := rekor_v1.RunConsistencyCheck(rekorClient, verifier, flags.LogInfoFile)
 			if err != nil {
-				fmt.Fprintf(os.Stderr, "error reading checkpoint: %v", err)
-				return
+				return nil, nil, err
 			}
-
-			checkpointEndIndex := rekor_v1.GetCheckpointIndex(logInfo, checkpoint)
-			config.EndIndex = &checkpointEndIndex
-		}
-
-		if *config.StartIndex >= *config.EndIndex {
-			fmt.Fprintf(os.Stderr, "start index %d must be strictly less than end index %d", *config.StartIndex, *config.EndIndex)
-			return
-		}
-
-		if identity.MonitoredValuesExist(monitoredValues) {
-			_, err = rekor_v1.IdentitySearch(ctx, *config.StartIndex, *config.EndIndex, rekorClient, monitoredValues, config.OutputIdentitiesFile, config.IdentityMetadataFile)
+			var prevCheckpoint cmd.Checkpoint
+			if prev != nil {
+				prevCheckpoint = prev
+			}
+			var curCheckpoint cmd.Checkpoint
+			if cur != nil {
+				curCheckpoint = cur
+			}
+			return prevCheckpoint, curCheckpoint, nil
+		},
+		GetStartIndexFn: func(prev, cur cmd.Checkpoint) *int {
+			checkpointStartIndex := rekor_v1.GetCheckpointIndex(cur.(*models.LogInfo), prev.(*util.SignedCheckpoint))
```

Comment:
I find this a bit confusing. Both `prev` and `cur` are `cmd.Checkpoint`, but they are not both checkpoints (`prev` is a checkpoint, `cur` is a `LogInfo`)